linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,455 @@
+
+            if (interlocked_add(&threadpool->quitting, 0) != 1)
+            {
+                int32_t existing_count = interlocked_add(&threadpool->task_array_count, 0);
+
+                // We have a new item
+                current_index = interlocked_add_64(&threadpool->consume_idx, 0) % existing_count;
+
+                // Is this index waiting for us to complete it
+                srw_lock_acquire_shared(threadpool->srw_lock);
+                TASK_RESULT curr_task_result = interlocked_compare_exchange(&threadpool->task_array[current_index].task_state, TASK_WORKING, TASK_WAITING);
+                srw_lock_release_shared(threadpool->srw_lock);
+                if (TASK_WAITING == curr_task_result)
+                {
+                    int64_t consume_pos = interlocked_compare_exchange_64(&threadpool->consume_idx, current_index+1, current_index);
+                    if (consume_pos == current_index)
+                    {
+                        srw_lock_acquire_shared(threadpool->srw_lock);
+                        THREADPOOL_TASK* temp_item = &threadpool->task_array[consume_pos];
+                        srw_lock_release_shared(threadpool->srw_lock);
+
+                        // Get the item and make sure it's valid
+                        if (temp_item != NULL)
+                        {
+                            temp_item->task_func(temp_item->task_param);
+                        }
+                        // I'm done with this item mark it as complete
+                        srw_lock_acquire_shared(threadpool->srw_lock);
+                        interlocked_exchange(&threadpool->task_array[consume_pos].task_state, TASK_NOT_USED);
+                        srw_lock_release_shared(threadpool->srw_lock);
+                    }
+                }
+            }
+        } while (interlocked_add(&threadpool->quitting, 0) != 1);
+        (void)interlocked_decrement(&threadpool->thread_count);
+    }
+}
+
+static int reallocate_threadpool_array(THREADPOOL* threadpool)
+{
+    int result;
+    {
+        // Lock the array here
+        srw_lock_acquire_exclusive(threadpool->srw_lock);
+
+        int32_t existing_count = interlocked_add(&threadpool->task_array_count, 0);
+
+        // Double the items
+        (void)interlocked_exchange(&threadpool->task_array_count, existing_count*2);
+
+        // Reallocate here
+        THREADPOOL_TASK* temp_array = realloc_2((void*)threadpool->task_array, existing_count*2, sizeof(THREADPOOL_TASK));
+        if (temp_array == NULL)
+        {
+            LogError("Failure realloc_2(threadpool->task_array: %p, threadpool->task_array_count: %" PRId32", sizeof(THREADPOOL_TASK): %zu", threadpool->task_array, existing_count*2, sizeof(THREADPOOL_TASK));
+            result = MU_FAILURE;
+        }
+        else
+        {
+            // Clear the newly allocated memory
+            memset(temp_array+existing_count, 0, existing_count*sizeof(THREADPOOL_TASK));
```

Comment:
>memset(temp_array+existing_count, 0, existing_count*sizeof(THREADPOOL_TASK)); [](http://example.com/codeflow?start=12&length=77)

I think we should explicitly initialize with a for loop and set the desired memobers to what they need to be set.
Because there are volatile_atomic members that should always be changed with interlocked only.